src/main/java/com/weather/discordweather/gateway/WeatherForecastGateway.java

```diff
@@ -30,13 +32,29 @@ public WeatherForecastGateway(
     openWeatherClient = wc;
   }
 
-  public Optional<WeatherForecast> getWeatherForecast(double lat, double lon) {
-    OneCallResponse openWeatherResponse = openWeatherClient.getWeather(lat, lon);
-    GeocodeResponse mapQuestResponse = mapQuestClient.reverseGeocode(lat, lon);
-    return WeatherForecastMapper.fromOpenWeatherMapAndMapQuest(
-        openWeatherResponse,
-        mapQuestResponse
-    );
+  public Optional<WeatherForecast> getWeatherForecast(Double lat, Double lon, String location) {
+    GeocodeResponse mapQuestResponse;
+    OneCallResponse openWeatherResponse = null;
+
+    if (!location.isEmpty()) {
+      mapQuestResponse = mapQuestClient.forwardGeocode(stripSpaces(location));
+      Optional<Coordinate> coordinates = extractCoordinates(mapQuestResponse);
+      if (coordinates.isPresent()) {
+        openWeatherResponse = openWeatherClient.getWeather(coordinates.get().lat(),
+            coordinates.get().lng());
+      }
+    } else {
+      mapQuestResponse = mapQuestClient.reverseGeocode(lat, lon);
+      openWeatherResponse = openWeatherClient.getWeather(lat, lon);
+    }
+
+    if (mapQuestResponse != null && openWeatherResponse != null) {
```

Comment:
Do these ever return `null`? I thought they'd throw an exception if there are issues?

Reply:
Hm yeah they never return null now that I think about it. I'll remove this check.

Reply:
Actually `openWeatherResponse` may never be set if we don't get coordinates. Though it seems like MapQuest always returns something even for gibberish.

Reply:
If `openWeatherResponse` will be `null` because of MapQuest returning something without coordinates, can we do `return Optional.empty()` at that point too since we know for sure we don't need to continue doing stuff?

Reply:
Ah yes good idea.